tutorials/javaFxPart5.md

```diff
@@ -0,0 +1,419 @@
+}
+```
+
+**dialog-box.css**
+```css
+.label {
+    -fx-background-color: linear-gradient(to bottom right, #00ffbf, #00ddff);
+    -fx-border-color: #d55e00 #009e73 #cc79a7 #0072b2;
+    -fx-border-width: 2px;
+    -fx-background-radius: 1em 1em 0 1em;
+    -fx-border-radius: 1em 1em 0 1em;
+}
+
+.reply-label {
+    -fx-background-radius: 1em 1em 1em 0;
+    -fx-border-radius: 1em 1em 1em 0;
+}
+
+#displayPicture {
+    /* Shadow effect on image. */
+    -fx-effect: dropshadow(gaussian, rgba(0, 0, 0, 0.2), 10, 0.5, 5, 5);
+
+    /* Change size of image. */
+    -fx-scale-x: 1;
+    -fx-scale-y: 1;
+
+    /* Rotate image clockwise by degrees. */
+    -fx-rotate: 0;
+}
+```
+
+Proceed with the following steps to link the css files to the FXML files through SceneBuilder.
+
+1. Open `MainWindow.fxml` in SceneBuilder, select the `AnchorPane`.
+
+2. On the right accordion pane, click on the plus sign under Stylesheets within Properties as shown below and choose
+   the previously created `src/main/resources/css/main.css` file.
+   ![Link CSS to MainWindow.fxml](images/javafx/LinkCSSToFXML1.png)
+
+3. If you don't see the updated view, try reopening the file with SceneBuilder again.
+   <box type="important" seamless bold>
+
+   **Remember to save the `.fxml` file after linking!**
+   </box>
+
+   ![Link CSS to MainWindow.fxml](images/javafx/LinkCSSToFXML2.png)
+
+
+4. Repeat steps 1-3 for `DialogBox.fxml` with `dialog-box.css`.
+   ![Link CSS to DialogBox.fxml](images/javafx/LinkCSSToFXML3.png)
+
+5. Save the FXML files and run the app to test it!
+
+   ![Duke with default CSS template](images/javafx/DefaultDukeCSSApp.png)
+
+### Update code for flipped text bubbles
+
+You may have noticed that the text bubbles are not facing the right direction for DaDuke.
+
+We created a style class specifically for this case:
+
+**dialog-box.css**
+```css
+.reply-label {
+    -fx-background-radius: 1em 1em 1em 0;
+    -fx-border-radius: 1em 1em 1em 0;
+}
+```
+
+We can add the `reply-label` style to the dialog box when we flip it.
+
+Update the following code:
+
+**DialogBox.java**
+```java
+ private void flip() {
+     // current code...
+     dialog.getStyleClass().add("reply-label");
+ }
+```
+
+## Padding and Margin
+
+In short, padding represents the space between the content and its border, while margin represents the invisible space
+around the border.
+
+![Padding and Margin](images/javafx/PaddingAndMargin.png)
+
+Let us play around with the padding and margin of the text bubble.
+
+Open `DialogBox.fxml` in SceneBuilder. Click on `Label`, and look at layout in the right accordion panel.
+
+![Padding and Margin Demo](images/javafx/PaddingAndMarginDemo.png)
+
+Play around with the padding and margin and find the parameters that is most suitable for you! To better view the
+effects of the margin, you may either
+- Run the app
+- Select both image and label to view the distance between them
+  ![View effects of margin](images/javafx/ViewEffectsOfMargin.png)
+
```

Comment:
Should we also include how to fix the text clipping problem? https://github.com/nus-cs2103-AY2324S2/forum/issues/208

Reply:
Yup, sounds good, will work on this.

Reply:
@damithc to follow up on this, since the issue is fixed by removing the `prefHeight` for VBox, is it possible to adjust that in the original code from the earlier parts instead?

Reply:
Sure, that works too.